File: app/models/email_confirmation.rb
# typed: strict
# frozen_string_literal: true

class EmailConfirmation < ApplicationRecord
  EXPIRES_IN = T.let(15.minutes, ActiveSupport::Duration)

  enum :event, {
    EmailConfirmationEvent::SignUp.serialize => 0,
    EmailConfirmationEvent::EmailUpdate.serialize => 0
  }, prefix: true

  scope :active, -> { where(succeeded_at: nil).where("started_at > ?", EXPIRES_IN.ago) }
  scope :succeeded, -> { where.not(succeeded_at: nil) }

  validates :email, presence: true
  validates :code, format: {with: /\A\d{6}\z/}, presence: true

  sig { returns(String) }
  def self.generate_code
    6.times.map { rand(10) }.join
  end

  sig { returns(T::Boolean) }
  def succeeded?
    !succeeded_at.nil?
  end

  sig { void }
  def success!
    update!(succeeded_at: Time.current) unless succeeded?

    nil
  end

  sig { params(locale: UserLocale).void }
  def send_mail!(locale:)
    EmailConfirmationMailer.email_confirmation(email_confirmation_id: id.not_nil!, locale: locale.serialize).deliver_later

    nil
  end
end

File: go/internal/model/email_confirmation.go
package model

import (
	"time"
)

// EmailConfirmationEventはメール確認イベントの種類を表す
type EmailConfirmationEvent int32

const (
	// EmailConfirmationEventSignUpはサインアップ時のメール確認
	EmailConfirmationEventSignUp EmailConfirmationEvent = 0
	// EmailConfirmationEventEmailUpdateはメールアドレス変更時の確認
	EmailConfirmationEventEmailUpdate EmailConfirmationEvent = 1
	// EmailConfirmationEventPasswordResetはパスワードリセット時の確認
	EmailConfirmationEventPasswordReset EmailConfirmationEvent = 2
)

// EmailConfirmationはメール確認のドメインモデル
type EmailConfirmation struct {
	ID          string
	Email       string
	Event       EmailConfirmationEvent
	Code        string
	StartedAt   time.Time
	SucceededAt *time.Time
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

// IsExpiredは確認コードの有効期限が切れているかを返す (15分)
func (e *EmailConfirmation) IsExpired() bool {
	return time.Since(e.StartedAt) > 15*time.Minute
}

// IsSucceededは確認が完了しているかを返す
func (e *EmailConfirmation) IsSucceeded() bool {
	return e.SucceededAt != nil
}
